lib/collection/response.js

```diff
@@ -96,12 +96,265 @@ var util = require('../util'),
+                   / "^" / "_" / "`" / "|" / "~"
+                   ; token except ( "*" / "'" / "%" )
+    *
+    * egHeader: attachment;filename*=utf-8''%E4%BD%A0%E5%A5%BD.txt
+    */
+    // eslint-disable-next-line max-len
+    encodedFileNameRegex = /;[ \t]*(?:filename\*)[ \t]*=[ \t]*([A-Za-z0-9!#$%&+\-^_`{}~]+)'.*'((?:%[0-9A-Fa-f]{2}|[A-Za-z0-9!#$&+.^_`|~-])+)[ \t]*/,
+
+    /**
+    * @private
+    *
+    * RegExp to match quoted-pair in RFC 2616
+    *
+    * quoted-pair = "\" CHAR
+    * CHAR        = <any US-ASCII character (octets 0 - 127)>
+    *
+    */
+    quotedPairRegex = /\\([ -~])/g,
+
+    /**
+     * @private
+     * Regex to match file base name and extension name
+     * Group 1 will be matched to the base name
+     * Group 2 will be matched to the extension name
+     */
+    fileNameSplitRegex = /^(.+?\.?)(?:\.([^.]*$)|$)/,
+
+    /**
+     * @private
+     * Regex to match all the hexadecimal number inside encoded string
+     */
+    hexCharMatchRegex = /%([0-9A-Fa-f]{2})/g,
+
+    /**
+     * @private
+     * Regex to match non-latin characters
+     */
+    nonLatinCharMatchRegex = /[^\x20-\x7e\xa0-\xff]/g,
+
+    /**
+     * Replaces non-latin characters with '?'
+     *
+     * @private
+     * @param {string} val - The string which needs to be checked
+     */
+    getLatin = function (val) {
+        return String(val).replace(nonLatinCharMatchRegex, '?');
+    },
+
+    /**
+     * Decodes the hexcode to charCode
+     *
+     * @private
+     * @param {string} str
+     * @param {string} hex
+     * @returns {string} String with decoded hexcode values
+     */
+    decodeHexcode = function (str, hex) {
+        return String.fromCharCode(parseInt(hex, 16));
+    },
+
+    /**
+     * Decodes the given filename with given charset
+     * The supported charset are iso-8859-1 and utf-8
+     *
+     * @private
+     * @param {string} encodedFileName
+     * @param {string} charset
+     * @returns {string} Returns the decoded filename
+     */
+    decodeFileName = function (encodedFileName, charset) {
+        if (!encodedFileName) {
+            return;
+        }
+        // Decodes the hexadecimal numbers to charCode
+        var binary = encodedFileName.replace(hexCharMatchRegex, decodeHexcode),
+            value;
+        switch (charset) {
+            case iso88591:
+                value = getLatin(binary);
+                break;
+            case utf8:
+                if (!supportsBuffer) {
+                    return;
+                }
+                value = (new Buffer(binary, BINARY)).toString(UTF8);
+                break;
+            default:
+                return;
+        }
+        return value;
+    },
+
+    /**
+     * Splits the file base name and extension name
+     *
+     * @private
+     * @param {string} fileName
+     * @returns {string} Returns the decoded filename
+     */
+    splitFileName = function (fileName) {
+        var fileNameSplitMatches;
```

Comment:
Just check once whether non-regex or multi-stage parsing having a waterfall of split and then loop can be more performant.